components/script/dom/htmlinputelement.rs

```diff
@@ -171,26 +273,34 @@ impl HTMLInputElement {
                            HTMLInputElementBinding::Wrap)
     }
 
-    pub fn type_(&self) -> Atom {
-        self.upcast::<Element>()
-            .get_attribute(&ns!(), &local_name!("type"))
-            .map_or_else(|| atom!(""), |a| a.value().as_atom().to_owned())
-    }
-
     // https://html.spec.whatwg.org/multipage/#dom-input-value
+    // https://html.spec.whatwg.org/multipage/#concept-input-apply
     fn value_mode(&self) -> ValueMode {
-        match self.input_type.get() {
-            InputType::InputSubmit |
-            InputType::InputReset |
-            InputType::InputButton |
-            InputType::InputImage => ValueMode::Default,
-            InputType::InputCheckbox |
-            InputType::InputRadio => ValueMode::DefaultOn,
-            InputType::InputPassword |
-            InputType::InputText => ValueMode::Value,
-            InputType::InputFile => ValueMode::Filename,
+        match self.input_type() {
+            InputType::Submit | InputType::Reset | InputType::Button
+            | InputType::Image | InputType::Hidden => {
+                ValueMode::Default
+            },
+
+            InputType::Checkbox | InputType::Radio => {
+                ValueMode::DefaultOn
+            },
+
+            InputType::Color | InputType::Date | InputType::Datetime
+            | InputType::DatetimeLocal | InputType::Email | InputType::Month
+            | InputType::Number | InputType::Password | InputType::Range
+            | InputType::Search | InputType::Tel | InputType::Text
+            | InputType::Time | InputType::Url | InputType::Week => {
+                ValueMode::Value
+            }
+
+            InputType::File => ValueMode::Filename,
         }
     }
+
+    pub fn input_type(&self) -> InputType {
```

Comment:
I don't actually feel like this is THAT much of an improvement, since this essentially only saves us from typing `.get`. In fact, I think this may even hurt performance since we have to go through another function call... but since you've already made the change everywhere to use `input_type()`, the least we can do to mitigate this is to add a `#[inline]` annotation above.

Reply:
Thanks, done. Note that there need to be some public way to get the `InputType` from outside of the crate, which is part of the reason that I added this function (the other part of the reason is indeed that I think it's slightly nicer to not have to repeat `input_type.get()` everywhere.